File: Algorithm/141. Linked List Cycle.py
# -*- coding: utf-8 -*-
# @Author: WuLC
# @Date:   2016-06-15 21:31:56
# @Last modified by:   WuLC
# @Last Modified time: 2016-06-28 14:28:20
# @Email: liangchaowu5@gmail.com


# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# method 1, modify the link list
class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        while head:
            if head == head.next: return True
            p = head.next 
            head.next = head
            head = p
        return False

# method 2, tow pointers
class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        p1 = p2 = head
        while p1 and p2:
            if p1.next:
                p1 = p1.next.next
            else:
                return False
            p2 = p2.next
            if p1 == p2:
                return True
        return False

File: Algorithm/Java/141. Linked List Cycle.java
/*
 * Created on Tue Mar 13 2018 22:44:21
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
 // two pointers
public class Solution 
{
    public boolean hasCycle(ListNode head) 
    {
        ListNode p1 = head, p2 = head;
        while (p1 != null && p2 != null && p2.next != null)
        {
            p1 = p1.next;
            p2 = p2.next.next;
            if (p1==p2) return true;
        }
        return false;
    }
}